File: src/lib/ocr.ts
import * as ImageManipulator from "expo-image-manipulator";
import * as ImagePicker from "expo-image-picker";
import { Platform } from "react-native";

import { config } from "./config";
import {
  ensureCreditsLoaded,
  getBuiltinModel,
  loadSelectedModelId,
  trySpendCredits,
} from "./credits";
import {
  buildChatCompletionsUrl,
  isCustomOcrConfigSet,
  loadOcrProviderConfig,
  requiresCustomOcrConfig,
} from "./ocrConfig";

const OCR_MAX_EDGE = 1600;
const OCR_JPEG_QUALITY = 0.82;
const OCR_IMAGE_PICKER: ImagePicker.ImagePickerOptions = {
  mediaTypes: ["images"],
  quality: 1,
  // Android system crop is square-only and flaky on some OEMs.
  allowsEditing: Platform.OS === "ios",
};

/** Disclaimer surfaced at OCR entry points so users know results may be off. */
export const OCR_DISCLAIMER = "AI 识图可能存在误差，请核对识别结果";

/**
 * Thrown when a premium built-in model is selected but the credit balance is
 * insufficient. The UI catches this to open the recharge flow instead of
 * showing a generic error toast.
 */
export class InsufficientCreditsError extends Error {
  constructor(public cost: number) {
    super("Credits 不足，请充值后使用高级识别");
    this.name = "InsufficientCreditsError";
  }
}

/** In-flight progress phases (header). Terminal copy lives in OCR_OUTCOME_MESSAGES. */
export type OcrProgressPhase =
  | "preparing_photo"
  | "preparing_album"
  | "compressing"
  | "recognizing";

export const OCR_PROGRESS_MESSAGES: Record<OcrProgressPhase, string> = {
  preparing_photo: "已拍摄，准备识别…",
  preparing_album: "已选图，准备识别…",
  compressing: "处理图片中…",
  recognizing: "识别中…",
};

export const OCR_OUTCOME_MESSAGES = {
  success: (count: number) => `已识别 ${count} 个单词`,
  empty: "未识别到英文单词，请换一张更清晰的图片再试",
  emptyUnparsed:
    "未能从识别结果中提取英文单词，请换一张更清晰的单词列表再试",
  failed: "识别失败",
} as const;

export type OcrUiState = {
  busy: boolean;
  message: string;
};

export const OCR_UI_IDLE: OcrUiState = { busy: false, message: "" };

function uriToBase64(uri: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.onload = () => {
      const reader = new FileReader();
      reader.onload = () => {
        const result = reader.result as string;
        const commaIndex = result.indexOf(",");
        resolve(commaIndex >= 0 ? result.slice(commaIndex + 1) : result);
      };
      reader.onerror = () => reject(new Error("读取图片失败"));
      reader.readAsDataURL(xhr.response);
    };
    xhr.onerror = () => reject(new Error("读取图片失败"));
    xhr.responseType = "blob";
    xhr.open("GET", uri);
    xhr.send();
  });
}

async function compressImageForOcr(
  uri: string,
): Promise<{ base64: string; mimeType: string }> {
  const resized = await ImageManipulator.manipulateAsync(
    uri,
    [
      {
        resize: {
          width: OCR_MAX_EDGE,
        },
      },
    ],
    {
      compress: OCR_JPEG_QUALITY,
      format: ImageManipulator.SaveFormat.JPEG,
    },
  );

  const base64 = await uriToBase64(resized.uri);
  return { base64, mimeType: "image/jpeg" };
}

export async function takePhoto(): Promise<string | null> {
  const permission = await ImagePicker.requestCameraPermissionsAsync();
  if (!permission.granted) {
    throw new Error("需要相机权限");
  }

  const result = await ImagePicker.launchCameraAsync(OCR_IMAGE_PICKER);

  if (result.canceled || !result.assets.length) return null;
  return result.assets[0]!.uri;
}

export async function pickFromAlbum(): Promise<string | null> {
  const permission = await ImagePicker.requestMediaLibraryPermissionsAsync();
  if (!permission.granted) {
    throw new Error("需要相册权限");
  }

  const result = await ImagePicker.launchImageLibraryAsync(OCR_IMAGE_PICKER);

  if (result.canceled || !result.assets.length) return null;
  return result.assets[0]!.uri;
}

interface OcrRequestConfig {
  baseUrl: string;
  apiKey: string;
  model: string;
  /** Credits to deduct after a successful call. 0 for free / BYOK. */
  premiumCost: number;
}

/**
 * Resolve the effective OCR provider for this run:
 *  1. A complete BYOK custom config wins (user's own key — never charges).
 *  2. Otherwise the selected built-in model is used against the app's embedded
 *     key. Premium built-in models require enough credits up-front.
 */
async function resolveOcrRequest(): Promise<OcrRequestConfig> {
  const custom = await loadOcrProviderConfig();
  if (isCustomOcrConfigSet(custom)) {
    return {
      baseUrl: custom!.baseUrl.trim(),
      apiKey: custom!.apiKey.trim(),
      model: custom!.model.trim(),
      premiumCost: 0,
    };
  }

  if (requiresCustomOcrConfig()) {
    throw new Error("请先在设置中配置 OCR 服务（Web 版需自备 API Key）");
  }
  if (!config.zhipuApiKey.trim()) {
    throw new Error("请先在设置中配置 OCR 服务");
  }

  const selectedId = await loadSelectedModelId();
  const selected = getBuiltinModel(selectedId);
  if (selected.tier === "premium") {
    const balance = await ensureCreditsLoaded();
    if (balance < selected.creditCost) {
      throw new InsufficientCreditsError(selected.creditCost);
    }
    return {
      baseUrl: config.zhipuBaseUrl,
      apiKey: config.zhipuApiKey,
      model: selected.model,
      premiumCost: selected.creditCost,
    };
  }

  return {
    baseUrl: config.zhipuBaseUrl,
    apiKey: config.zhipuApiKey,
    model: selected.model,
    premiumCost: 0,
  };
}

export async function ocrWordsFromImage(
  imageUri: string,
  onProgress?: (phase: OcrProgressPhase) => void,
): Promise<{ words: string[]; rawText: string }> {
  onProgress?.("compressing");
  const { base64, mimeType } = await compressImageForOcr(imageUri);
  const dataUrl = `data:${mimeType};base64,${base64}`;

  onProgress?.("recognizing");

  const { baseUrl, apiKey, model, premiumCost } = await resolveOcrRequest();
  const endpoint = buildChatCompletionsUrl(baseUrl);

  let response: Response;
  try {
    response = await fetch(endpoint, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model,
        temperature: 0.1,
        messages: [
          {
            role: "user",
            content: [
              {
                type: "image_url",
                image_url: { url: dataUrl },
              },
              {
                type: "text",
                text: [
                  "这是一张包含英文单词列表的图片。",
                  "请识别图中所有英文单词或词组。",
                  "如果单词旁边标注了词性和中文释义，请一并提取，每行格式：单词 | 词性 | 中文释义",
                  "如果图中没有词性或释义信息，只输出单词本身。",
                  "像 actor / actress 这样的斜杠词组应作为一整行输出，不要拆开。",
                  "不要用逗号连接、不要编号、不要输出其他标点或解释。",
                ].join(""),
              },
            ],
          },
        ],
      }),
    });
  } catch {
    throw new Error("网络请求失败，请检查网络后重试");
  }

  if (!response.ok) {
    // Read the error body BEFORE throwing: the old shape threw inside its
    // own try/catch, which swallowed the detailed message and always
    // surfaced the generic fallback to the user.
    const detail = await response.text().catch(() => "");
    throw new Error(detail ? `视觉识别失败: ${detail}` : "视觉识别服务异常");
  }

  // Charge credits only after a successful API response so failed/errored
  // calls don't consume the user's balance. OcrSection's runningRef
  // prevents a double-tap from starting two spends.
  if (premiumCost > 0) {
    await trySpendCredits(premiumCost);
  }

  const payload = (await response.json()) as {
    choices?: Array<{ message?: { content?: string } }>;
  };
  const rawText = payload.choices?.[0]?.message?.content?.trim() ?? "";
  const words = extractWordsFromOcrText(rawText);

  return { words, rawText };
}

/**
 * Verify a candidate OCR provider config by sending a minimal text-only
 * chat/completions request. Throws with a descriptive message on failure.
 * Used by the OCR settings panel before saving.
 */
export async function testOcrConfig(cfg: {
  baseUrl: string;
  apiKey: string;
  model: string;
}): Promise<void> {
  const endpoint = buildChatCompletionsUrl(cfg.baseUrl);
  let response: Response;
  try {
    response = await fetch(endpoint, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${cfg.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: cfg.model,
        temperature: 0,
        max_tokens: 1,
        messages: [{ role: "user", content: "hi" }],
      }),
    });
  } catch {
    throw new Error("网络请求失败，请检查 URL 与网络");
  }

  if (!response.ok) {
    let detail = "";
    try {
      const text = await response.text();
      try {
        const j = JSON.parse(text) as { error?: { message?: string } };
        detail = j.error?.message ?? text;
      } catch {
        detail = text;
      }
    } catch {
      // ignore
    }
    const hint = detail ? `: ${detail.slice(0, 160)}` : "";
    if (response.status === 401 || response.status === 403) {
      throw new Error(`认证失败（${response.status}）${hint}`);
    }
    if (response.status === 404) {
      throw new Error(`未找到接口（404），请检查 URL${hint}`);
    }
    throw new Error(`请求失败（${response.status}）${hint}`);
  }
}

/**
 * Real dictation phrases are almost never longer than this many space-separated
 * tokens (e.g. "ice cream", "look forward to", "actor / actress"). Longer runs
 * are treated as a failed one-line dump from the vision model.
 */
const MAX_PHRASE_TOKENS = 4;

/**
 * Strip list markers, wrapping quotes, and trailing punctuation from a token.
 */
function cleanToken(s: string): string {
  return s
    .replace(/^[\d.)\-•*、]+\s*/, "")
    .replace(/^["'`“”‘’]+|["'`“”‘’]+$/g, "")
    .replace(/[.。:：]+$/g, "")
    .trim();
}

const WORD_RE = /^[a-zA-Z][a-zA-Z'/\-\s]*$/;

/**
 * The leading English token run of `tokens` — the headword phrase. Empty
 * when the row does not begin with English, so a row that opens with
 * Chinese (or pure 音标) is not an English entry at all. A lone `/` keeps
 * the run going only while it is flanked by word tokens, so spaced slash
 * phrases (`actor / actress`) survive whole — the format the OCR prompt
 * asks for — while a trailing `/` or annotation tokens still end the run.
 */
function leadingEnglishRun(tokens: string[]): string[] {
  const run: string[] = [];
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i];
    if (WORD_RE.test(token)) {
      run.push(token);
    } else if (
      token === "/" &&
      run.length > 0 &&
      i + 1 < tokens.length &&
      WORD_RE.test(tokens[i + 1])
    ) {
      run.push(token);
    } else {
      break;
    }
  }
  return run;
}

/**
 * English tokens that start a cleaned token run — the headword phrase.
 * Null when the run does not start with English, so a row that begins
 * with Chinese (or pure 音标) is not an English entry at all.
 */
function englishPhrase(tokenRun: string): string | null {
  const phrase = leadingEnglishRun(tokenRun.split(/\s+/).filter(Boolean)).slice(
    0,
    MAX_PHRASE_TOKENS,
  );
  return phrase.length > 0 ? phrase.join(" ") : null;
}

/**
 * Candidate words from one cleaned comma-part. Pure-English runs keep the
 * original behavior: the whole phrase up to MAX_PHRASE_TOKENS tokens,
 * longer dumps flattened into single tokens. Mixed runs — the word
 * followed by 音标/词性/中文 (the vision model echoing a textbook row
 * instead of the `word | pos | meaning` format) — yield the leading
 * English phrase; the trailing annotations are noise and never become
 * entries.
 */
function englishCandidates(candidate: string): string[] {
  const tokens = candidate.split(/\s+/).filter(Boolean);
  const leading = leadingEnglishRun(tokens);
  if (leading.length === 0) return [];
  if (leading.length === tokens.length) {
    if (leading.length <= MAX_PHRASE_TOKENS) return [candidate];
    // Over-long pure dump flattened to single tokens; lone "/" separators
    // are noise and never become entries.
    return tokens.filter((t) => WORD_RE.test(t));
  }
  return leading.length <= MAX_PHRASE_TOKENS
    ? [leading.join(" ")]
    : leading.filter((t) => WORD_RE.test(t));
}
/**
 * Vision models often ignore "one per line" and return comma- or space-separated
 * lists. This function normalizes the output into a clean word list.
 *
 * Supports enriched entries with `|`-delimited pos/meaning:
 *   `apple | n. | 苹果` — the leading English phrase of the word part is kept
 * (a 音标 run trailing the headword is dropped); the meta is preserved. Plain
 * entries (no `|`) are comma-split; each part yields only its leading English
 * phrase — the whole phrase (≤ MAX_PHRASE_TOKENS tokens), individual tokens
 * of an over-long dump, or the headword of a row that also carries
 * 音标/词性/中文. Chinese text never becomes an entry.
 */
export function extractWordsFromOcrText(rawText: string): string[] {
  const cleaned = rawText
    .replace(/```[\s\S]*?```/g, (block) =>
      block.replace(/^```\w*\n?/, "").replace(/\n?```$/, ""),
    )
    .replace(/\r\n?/g, "\n");

  // Split by newlines first to keep enriched entries intact
  const lines = cleaned
    .split(/\n/)
    .map((l) => l.trim())
    .filter((l) => l.length > 0);

  const seen = new Set<string>();
  const words: string[] = [];

  for (const line of lines) {
    const pipeIdx = line.indexOf("|");

    if (pipeIdx !== -1) {
      // Enriched entry: salvage the leading English phrase of the word
      // part (a 音标 run trailing the headword is dropped), keep the meta.
      const wordPart = englishPhrase(cleanToken(line.slice(0, pipeIdx)));
      if (!wordPart) continue;

      const key = wordPart.toLowerCase();
      if (seen.has(key)) continue;
      seen.add(key);

      const metaParts = line
        .slice(pipeIdx + 1)
        .split("|")
        .map((s) => s.trim())
        .filter((s) => s.length > 0);

      // Reassemble with consistent spacing
      words.push([wordPart, ...metaParts].join(" | "));
    } else {
      // Plain line: split by commas/semicolons (the vision model may
      // ignore one-per-line), then keep the headword of each part.
      const candidates = line
        .split(/[,，;；、]+/)
        .map((p) => cleanToken(p))
        .filter(Boolean)
        .flatMap(englishCandidates);

      for (const word of candidates) {
        const key = word.toLowerCase();
        if (seen.has(key)) continue;
        seen.add(key);
        words.push(word);
      }
    }
  }

  return words;
}

File: lib/services/ocr.dart
import 'dart:convert';

import 'package:flutter_image_compress/flutter_image_compress.dart';
import 'package:http/http.dart' as http;
import 'package:image_picker/image_picker.dart';

import 'config.dart';
import 'credits.dart';
import 'ocr_config.dart';

/// 拍照识词。对应 RN 版 src/lib/ocr.ts。
///
/// - expo-image-picker      → image_picker
/// - expo-image-manipulator → flutter_image_compress
/// - fetch                  → package:http
const int _ocrMaxEdge = 1600;
const int _ocrJpegQuality = 82;

/// OCR 入口处统一展示的免责声明，让用户知道结果可能有误。
const String kOcrDisclaimer = 'AI 识图可能存在误差，请核对识别结果';

/// 选中高级内置模型但余额不足时抛出。
/// UI 捕获它来打开充值流程，而不是弹一个通用错误 toast。
class InsufficientCreditsError implements Exception {
  const InsufficientCreditsError(this.cost);

  final int cost;

  @override
  String toString() => 'Credits 不足，请充值后使用高级识别';
}

/// 进行中的阶段（顶栏）。终态文案在 OcrOutcomeMessages。
enum OcrProgressPhase {
  preparingPhoto,
  preparingAlbum,
  compressing,
  recognizing
}

const Map<OcrProgressPhase, String> kOcrProgressMessages = {
  OcrProgressPhase.preparingPhoto: '已拍摄，准备识别…',
  OcrProgressPhase.preparingAlbum: '已选图，准备识别…',
  OcrProgressPhase.compressing: '处理图片中…',
  OcrProgressPhase.recognizing: '识别中…',
};

class OcrOutcomeMessages {
  const OcrOutcomeMessages._();

  static String success(int count) => '已识别 $count 个单词';
  static const String empty = '未识别到英文单词，请换一张更清晰的图片再试';
  static const String emptyUnparsed = '未能从识别结果中提取英文单词，请换一张更清晰的单词列表再试';
  static const String failed = '识别失败';
}

class OcrUiState {
  const OcrUiState({required this.busy, required this.message});

  final bool busy;
  final String message;

  static const OcrUiState idle = OcrUiState(busy: false, message: '');
}

final ImagePicker _picker = ImagePicker();

Future<String?> takePhoto() async {
  final file = await _picker.pickImage(
    source: ImageSource.camera,
    imageQuality: 100,
  );
  return file?.path;
}

Future<String?> pickFromAlbum() async {
  final file = await _picker.pickImage(
    source: ImageSource.gallery,
    imageQuality: 100,
  );
  return file?.path;
}

class _CompressedImage {
  const _CompressedImage(this.base64, this.mimeType);

  final String base64;
  final String mimeType;
}

Future<_CompressedImage> _compressImageForOcr(String path) async {
  final bytes = await FlutterImageCompress.compressWithFile(
    path,
    minWidth: _ocrMaxEdge,
    minHeight: _ocrMaxEdge,
    quality: _ocrJpegQuality,
    format: CompressFormat.jpeg,
  );

  if (bytes == null) {
    throw Exception('读取图片失败');
  }
  return _CompressedImage(base64Encode(bytes), 'image/jpeg');
}

class _OcrRequestConfig {
  const _OcrRequestConfig({
    required this.baseUrl,
    required this.apiKey,
    required this.model,
    required this.premiumCost,
  });

  final String baseUrl;
  final String apiKey;
  final String model;

  /// 成功后要扣的 credits，免费/BYOK 为 0。
  final int premiumCost;
}

/// 决定这一次识别用哪个服务：
///  1. 完整的 BYOK 自定义配置优先（用户自己的 key —— 永不扣费）。
///  2. 否则用选中的内置模型 + App 内置密钥。高级内置模型需要预先有余额。
Future<_OcrRequestConfig> _resolveOcrRequest() async {
  final custom = await loadOcrProviderConfig();
  if (isCustomOcrConfigSet(custom)) {
    return _OcrRequestConfig(
      baseUrl: custom!.baseUrl.trim(),
      apiKey: custom.apiKey.trim(),
      model: custom.model.trim(),
      premiumCost: 0,
    );
  }

  if (requiresCustomOcrConfig()) {
    throw Exception('请先在设置中配置 OCR 服务（Web 版需自备 API Key）');
  }
  if (AppConfig.zhipuApiKey.trim().isEmpty) {
    throw Exception('请先在设置中配置 OCR 服务');
  }

  final selected = getBuiltinModel(await loadSelectedModelId());
  if (selected.tier == ModelTier.premium) {
    final balance = await ensureCreditsLoaded();
    if (balance < selected.creditCost) {
      throw InsufficientCreditsError(selected.creditCost);
    }
    return _OcrRequestConfig(
      baseUrl: AppConfig.zhipuBaseUrl,
      apiKey: AppConfig.zhipuApiKey,
      model: selected.model,
      premiumCost: selected.creditCost,
    );
  }

  return _OcrRequestConfig(
    baseUrl: AppConfig.zhipuBaseUrl,
    apiKey: AppConfig.zhipuApiKey,
    model: selected.model,
    premiumCost: 0,
  );
}

class OcrResult {
  const OcrResult({required this.words, required this.rawText});

  final List<String> words;
  final String rawText;
}

const String _ocrPrompt = '这是一张包含英文单词列表的图片。'
    '请识别图中所有英文单词或词组。'
    '如果单词旁边标注了词性和中文释义，请一并提取，每行格式：单词 | 词性 | 中文释义'
    '如果图中没有词性或释义信息，只输出单词本身。'
    '像 actor / actress 这样的斜杠词组应作为一整行输出，不要拆开。'
    '不要用逗号连接、不要编号、不要输出其他标点或解释。';

Future<OcrResult> ocrWordsFromImage(
  String imagePath, {
  void Function(OcrProgressPhase phase)? onProgress,
}) async {
  onProgress?.call(OcrProgressPhase.compressing);
  final image = await _compressImageForOcr(imagePath);
  final dataUrl = 'data:${image.mimeType};base64,${image.base64}';

  onProgress?.call(OcrProgressPhase.recognizing);

  final cfg = await _resolveOcrRequest();
  final endpoint = buildChatCompletionsUrl(cfg.baseUrl);

  http.Response response;
  try {
    response = await http.post(
      Uri.parse(endpoint),
      headers: {
        'Authorization': 'Bearer ${cfg.apiKey}',
        'Content-Type': 'application/json',
      },
      body: json.encode({
        'model': cfg.model,
        'temperature': 0.1,
        'messages': [
          {
            'role': 'user',
            'content': [
              {
                'type': 'image_url',
                'image_url': {'url': dataUrl},
              },
              {'type': 'text', 'text': _ocrPrompt},
            ],
          },
        ],
      }),
    );
  } catch (_) {
    throw Exception('网络请求失败，请检查网络后重试');
  }

  if (response.statusCode < 200 || response.statusCode >= 300) {
    final detail = utf8.decode(response.bodyBytes, allowMalformed: true);
    throw Exception(detail.isNotEmpty ? '视觉识别失败: $detail' : '视觉识别服务异常');
  }

  // 只在 API 返回成功之后才扣 credits，失败/异常不消耗余额。
  // 单飞（ocrBusy）保证不会并发扣费。
  if (cfg.premiumCost > 0) {
    await trySpendCredits(cfg.premiumCost);
  }

  final payload =
      json.decode(utf8.decode(response.bodyBytes)) as Map<String, dynamic>;
  final choices = payload['choices'];
  String rawText = '';
  if (choices is List && choices.isNotEmpty) {
    final message = (choices.first as Map)['message'];
    if (message is Map && message['content'] is String) {
      rawText = (message['content'] as String).trim();
    }
  }

  return OcrResult(words: extractWordsFromOcrText(rawText), rawText: rawText);
}

/// 用一个最小的纯文本 chat/completions 请求验证候选 OCR 配置。
/// 失败时抛出带描述的异常。OCR 设置面板保存前会调用它。
Future<void> testOcrConfig({
  required String baseUrl,
  required String apiKey,
  required String model,
}) async {
  final endpoint = buildChatCompletionsUrl(baseUrl);

  http.Response response;
  try {
    response = await http.post(
      Uri.parse(endpoint),
      headers: {
        'Authorization': 'Bearer $apiKey',
        'Content-Type': 'application/json',
      },
      body: json.encode({
        'model': model,
        'temperature': 0,
        'max_tokens': 1,
        'messages': [
          {'role': 'user', 'content': 'hi'},
        ],
      }),
    );
  } catch (_) {
    throw Exception('网络请求失败，请检查 URL 与网络');
  }

  final status = response.statusCode;
  if (status >= 200 && status < 300) return;

  var detail = '';
  try {
    final text = utf8.decode(response.bodyBytes, allowMalformed: true);
    try {
      final j = json.decode(text);
      if (j is Map && j['error'] is Map && j['error']['message'] is String) {
        detail = j['error']['message'] as String;
      } else {
        detail = text;
      }
    } catch (_) {
      detail = text;
    }
  } catch (_) {
    // 忽略
  }

  final hint = detail.isNotEmpty
      ? ': ${detail.length > 160 ? detail.substring(0, 160) : detail}'
      : '';
  if (status == 401 || status == 403) {
    throw Exception('认证失败（$status）$hint');
  }
  if (status == 404) {
    throw Exception('未找到接口（404），请检查 URL$hint');
  }
  throw Exception('请求失败（$status）$hint');
}

/// 真实听写词组几乎不会超过这么多个空格分隔的 token
/// （如 "ice cream"、"look forward to"、"actor / actress"）。
/// 更长的串按「视觉模型把结果堆成一行」处理。
const int _maxPhraseTokens = 4;

/// 去掉列表标记、外层引号和尾部标点。
String _cleanToken(String s) {
  return s
      .replaceFirst(RegExp(r'^[\d.)\-•*、]+\s*'), '')
      .replaceAll(RegExp(r'''^["'`“”‘’]+|["'`“”‘’]+$'''), '')
      .replaceAll(RegExp(r'[.。:：]+$'), '')
      .trim();
}

final RegExp _wordRe = RegExp(r"^[a-zA-Z][a-zA-Z'/\-\s]*$");

/// 视觉模型经常无视「每行一个」，返回逗号或空格分隔的列表。
/// 这个函数把输出归一化成干净的单词表。
///
/// 支持带 `|` 词性/释义的条目：
///   `apple | n. | 苹果` —— 校验单词部分，保留元数据。
/// 纯条目（无 `|`）沿用原来的逗号切分 + token 展平逻辑。
List<String> extractWordsFromOcrText(String rawText) {
  final cleaned = rawText
      .replaceAllMapped(
        RegExp(r'```[\s\S]*?```'),
        (m) => m
            .group(0)!
            .replaceFirst(RegExp(r'^```\w*\n?'), '')
            .replaceFirst(RegExp(r'\n?```$'), ''),
      )
      .replaceAll(RegExp(r'\r\n?'), '\n');

  // 先按换行切，保住带释义的整行条目
  final lines =
      cleaned.split('\n').map((l) => l.trim()).where((l) => l.isNotEmpty);

  final seen = <String>{};
  final words = <String>[];

  for (final line in lines) {
    final pipeIdx = line.indexOf('|');

    if (pipeIdx != -1) {
      // 带释义的条目：清洗并校验单词部分，元数据原样保留
      final rawWord = line.substring(0, pipeIdx);
      final metaParts = line
          .substring(pipeIdx + 1)
          .split('|')
          .map((s) => s.trim())
          .where((s) => s.isNotEmpty)
          .toList();

      final wordPart = _cleanToken(rawWord);
      if (wordPart.isEmpty || !_wordRe.hasMatch(wordPart)) continue;

      final key = wordPart.toLowerCase();
      if (!seen.add(key)) continue;

      // 用统一的间距重新拼装
      words.add([wordPart, ...metaParts].join(' | '));
    } else {
      // 纯单词：先按逗号/分号切（视觉模型可能忽略每行一个），
      // 再把过长的串展平成单个 token
      final candidates = <String>[];
      for (final part in line.split(RegExp(r'[,，;；、]+'))) {
        final candidate = _cleanToken(part);
        if (candidate.isEmpty) continue;
        final tokens =
            candidate.split(RegExp(r'\s+')).where((t) => t.isNotEmpty).toList();
        if (tokens.length <= _maxPhraseTokens) {
          candidates.add(candidate);
        } else {
          candidates.addAll(tokens);
        }
      }

      for (final word in candidates) {
        if (!_wordRe.hasMatch(word)) continue;
        final key = word.toLowerCase();
        if (!seen.add(key)) continue;
        words.add(word);
      }
    }
  }

  return words;
}
